---
title: "Eco Geno 3/20/17"
author: "R Kirsten Tyler"
date: "3/20/2017"
output: html_document
---

Today: 

**Finishing allele frequency and nucleotide diversity**

- estimate allele frequencies between H and S >> f(H) - f(S)

- Fst between H vs. S
    >>output to local machine (my laptop)
    >>plot in R
    
- Estimate synonymous nucleotide diversity, non-synonymous diversity, and the ratio of them
    >>output to local machine
    >>compare to Romiguier


*In Terminal, let's use VSFtools to filter genotypes and gzip that ish*    
```{}
##PATH TO THE FINAL VCF DATA (all 24 INDS): 

/data/project_data/snps/reads2snps/SSW_by24inds.txt.vcf.gz
Use VCFTools to filter genotypes and save in your home directory in gzipped format:
 
$ cd /data/project_data/snps/reads2snps
$ vcftools --gzvcf SSW_by24inds.txt.vcf.gz --min-alleles 2 --max-alleles 2 --maf 0.02 --max-missing 0.8 --recode --out ~/SSW_all_biallelic.MAF0.02.Miss0.8  
$ cd ~/
$ gzip SSW_all_biallelic.MAF0.02.Miss0.8.recode.vcf
```

**Let's practice parcing out different groups!!**
We will re-create our text files containing the individual ID's for just Sick (and separately, just Healthy) individuals

Use grep to pull out individuals that we want
Use cut to grab the first column (individual sample IDs)

```{}
##grab metadata from here: 
/data/project_data/snps/reads2snps/ssw_healthloc.txt

##then use grep and cut to get what we WANT
$ cd /data/project_data/snps/reads2snps/
$ grep "HH" ssw_healthloc.txt | cut -f1 >~/H_SampleIDs.txt

##let's check out what we have here. what did we grep and cut? wc = word count
$ wc H_SampleIDs.txt 
##8  8 24 H_SampleIDs.txt 

##more grepping and cutting
cd /data/project_data/snps/reads2snps/
$ grep "HH" ssw_healthloc.txt | cut -f1 >~/H_SampleIDs.txt
$ grep "HS\|SS" ssw_healthloc.txt | cut -f1 >~/S_SampleIDs.txt
```

**Estimate the frequency of alleles at each SNP**

```{}
##for healthy:
vcftools --gzvcf SSW_all_biallelic.MAF0.02.Miss0.8.recode.vcf.gz --freq2 --keep H_SampleIDs.txt --out H_AlleleFreqs

##and for sick: 
$ vcftools --gzvcf SSW_all_biallelic.MAF0.02.Miss0.8.recode.vcf.gz --freq2 --keep S_SampleIDs.txt --out S_AlleleFreqs
```

**Let's calculate Wright's Fst betweeen H and S groups = generate Fst values**

```{}
##we are gtelling it to calc fst between the two groups wier...H_sample and weir...S_sample and then what we want to call the file (--out HvS...)

vcftools --gzvcf SSW_all_biallelic.MAF0.02.Miss0.8.recode.vcf.gz --weir-fst-pop H_SampleIDs.txt --weir-fst-pop S_SampleIDs.txt --out HvS_Fst
```

**Let's download these files into our local machines**
Open up a new terminal and transfer the files into a folder on to our computer
```{}
##here we are in a new terminal (not signed in as rtyler1pbio381 bla bla)

ip0af52e1e:~ kirstensmac$ cd /Users/kirstensmac/Documents/Ecological_Genomics_PBIO381/Pop_Gen 
ip0af52e1e:Pop_Gen kirstensmac$ scp rtyler1@pbio381.uvm.edu:~/HvS_Fst.weir.fst .

rtyler1@pbio381.uvm.edu's password: 
HvS_Fst.weir.fst                              100%  114KB 114.2KB/s   00:00    
ip0af52e1e:Pop_Gen kirstensmac$ scp rtyler1@pbio381.uvm.edu:~/*.frq .

rtyler1@pbio381.uvm.edu's password: 
H_AlleleFreqs.frq                             100%  425KB 424.9KB/s   00:00    
ip0af52e1e:Pop_Gen kirstensmac$ scp rtyler1@pbio381.uvm.edu:~/S_AlleleFreqs.frq .

rtyler1@pbio381.uvm.edu's password: 
S_AlleleFreqs.frq                             100%  448KB 448.4KB/s   00:00    

```

**fix the files so R will like them**

**Let's look at the data in R**

```{r}
setwd("~/Documents/Ecological_Genomics_PBIO381/Pop_Gen")

list.files()

H_freq <- read.table("H_AlleleFreqs.frq", header = T)
S_freq <- read.table("S_AlleleFreqs.frq", header = T)

str(H_freq)

All_freq <- merge(H_freq, S_freq, by = c("CHROM", "POS"))
str(All_freq) #shows the structure of the data
head(All_freq)

All_freq$diff <- (All_freq$H_ALT - All_freq$S_ALT)
hist(All_freq$diff, breaks=50, col="red", main = "Allele frequency difference (H-S)")

fst <- read.table("HvS_Fst.weir.fst", header=T)
All_freq.fst <- merge(All_freq, fst, by=c("CHROM", "POS"))

plot(All_freq.fst$diff, All_freq.fst$WEIR_AND_COCKERHAM_FST, xlab="Allele frequency difference (H-S)", ylab="Fst", main="Healthy vs. Sick SNP divergence")

# we are seeing as the allele frequecy increases, fst increases. points are SNPs. fst is the proportion of allele frequency variation that is distributed bw groups. for allele freqencies, positive numbers are H, negative numbers are S. potentially an immune response??

#once we can visualize our results like this and see the potential outliers (which SNPs are most divergent?? which SNPs are being effected by disease?), we can take  look at the outlying SNPs...look at the gene ID...and eventually look at what the funtion is

#how can we find which of those genes have the greatest divergence?? the which command!!!

All_freq[which(All_freq.fst$WEIR_AND_COCKERHAM_FST>0.2),]
#they seem to be on different transcripts

#once we have the annotation info, we can merge that file in R and ask what are the funtions of these transcripts?

```


***March 22, 2017 - continuing on with this tutorial:***       







Let's get the estimates for nucleotide diversity in synon and nonsynon sites. It's also going to give us confidence intervals with bootstraps

```{}
##first we are looking at the data: 
[rtyler1@pbio381 reads2snps]$ head SSW_by24inds.txt.fas
>TRINITY_DN47185_c0_g1_TRINITY_DN47185_c0_g1_i2_g.24943_m.24943|sp|03|Allele_1

#now let's compute nucleotide diversity!! This takes a while so we'll do it in "screen"

[rtyler1@pbio381 reads2snps]$ screen (to do stuff in the background)
[rtyler1@pbio381 reads2snps]$ ### ctrl A then D to dettach
[rtyler1@pbio381 reads2snps]$ screen -r ###to reattach

[rtyler1@pbio381 reads2snps]$/data/popgen/dNdSpiNpiS_1.0 -alignment_file=SSW_by24inds.txt.fas -ingroup=sp -out=~/dNdSpiNpiS_output
```

While we wait for that, let's look at the summary from Steve's previous smaller VCF file run on just one sample per indiv. Here's that output: 




#                              Biological Summary                              #
################################################################################

Selected ingroup species: sp

Number of analyzed individual: 24 (from 1 population(s))

Total number of contig used for sequence analysis: 1113

Total number of SNPs: 5040

  - Biallelic: 4991
  - Triallelic: 49
  - Quadriallelic: 0

Fit: ##related to fst but takes into account any inbreeding that may have been happening in the data. high values mean more homozygosity than you would expect under random mating. Vary from -1 to 1

Average Fit: -0.0507419 [-0.06817; -0.031933] ## these look to be randomly mating and not inbreeding becasue of the negative numbers
(Fit calculated in 902 contigs)

Weir & Cockerham Fit (Evolution 1984):

Average Weir & Cockerham Fit: 0.00703754 [-0.017669; 0.032047]
(Fit calculated in 902 contigs)

piN/piS ratio: nucleotide diversity

Average piS in focal species: 0.00585312 [0.005172; 0.006598] ##mean of synonymous sites, then conf interval
Average piN in focal species: 0.00154546 [0.00133; 0.001782] ##mean of nonsynon sites - purifying selection is happening
Average piN / average piS: 0.264041 [0.223914; 0.310575] ##index (ratio...bw 0 and 1) of how effective selection is at clearing out deleterious mutations (purifying selection). the higher the ratio, more deleterious mtuations are occuring and selection isn't as effective at eliminating non-synon mutations that are deleterious. We will see a high ratio in small pops that are not diverse. high in long lived vertebrates but low in huge pop sizes like bacteria and invertebrates etc
(piS and piN calculated in 902 contigs of average length 50)

***so let's compare this info to the Romiguier data to see how these sea stars fall!!***

```{r}
# Set your working directory to where you downloaded your file:
setwd("~/Documents/Ecological_Genomics_PBIO381")

# List the files in this directory
list.files()

# Read in the Romiguier data:
Rom <- read.csv("Romiguier_nature13685-s3.csv", header=T)

# Import OK?
str(Rom) 
head(Rom)

# Looks good!
# Now let's look at how the strength of purifying selection (piN/piS) compares to the size of Ne (piS). We'll plot these on a log scale to linearize the relationship.
plot(log(Rom$piS), log(Rom$piNpiS), pch=21, bg="blue", xlab="log Synonymous Nucleotide Diversity (piS)", ylab="log Ratio of Nonysn to Syn Diversity (piN/piS)", main="Purifying Selection vs. Effective Population Size")

# Now let's add our SSW points to the existing plot and give them a different symbol
points(log(0.00585312), log(0.264041), pch=24, cex=1.5, bg="red") 

# We can also add a regression line to the plot to see how far off the SSW estimates are from expectation
reg <- lm(log(Rom$piNpiS) ~ log(Rom$piS)) # Fits a linear regression
abline(reg) # adds the regression line to the plot

# It would be useful to highlight the other echinoderms in the dataset...do our seastars behave similarly?
echino <- Rom[which(Rom$Phylum=="Echinodermata"),] # subsets the data
points(log(echino$piS), log(echino$piNpiS), pch=21, bg="red") # adds the points

# Lastly, let's add a legend:
legend("bottomleft", cex=1, legend=c("Metazoans", "Echinoderms", "P. ochraceus"), pch=c(21,21,24), col=c("blue", "red", "red"))

# Pisaster seems to be in a group with other echinoderms that have relaxed purifying selection (high piN/piS), given their Ne...Interesting! Can we hypothesize why this might be?

echino #compare ours to the other echinos

#now let's estimate effective population size based on previous literature and our data: 
#this assumes that generation time is the same as the generation time used for the previous lit
.00585/(4*4*(10^-9))
```




















