---
title: "Eco Gen0 3/6/17"
author: "R Kirsten Tyler"
date: "3/6/2017"
output: html_document
---

###VCF TOOLS###

We are interesterd in:   

 - The position of where SNPs are located within a contig   
 - Which alleles are present in that location? A T C G   
 - Are there more than 2 alleles per location?   
 - How many reads were observed for each allele?   
 - What's the quality of the genotype?   
 
 
 Header
 Column names: 
    Chrom = Transcript ID (because we don't have chromosome numbers here)
    Ref = allele present in the assembly that we mapped everything to
    Alt = whatever other allleles were in the sample
    Qual = how confident are we that a SNP is located there
    Info = diff kinds of info: depth, how many reads are covering it, genotype quality, evidence of paralogy
    
    Then come the data!!

##Our workhorse for this SNP data is called VCF TOOLS##
We want to get some info and filter SNPs based on depth, genotype quality, allele frequency, calc statistics, and it allows us to parset the total data file into certain SNPs (ie just the SNPs that show differential expression in two groups; or just a subset of indiv in a file to calc diversity)


#In Terminal#

```{}
[rtyler1@pbio381 ~]$ vcftools
[rtyler1@pbio381 ~]$ cd /data/project_data/snps/reads2snps/
[rtyler1@pbio381 reads2snps]$ ll

```

VIM
Looking at the head of the data, we have SNPs on the y and individual samples on the x
applying two filters: depth filter of 10 reads for each genotype call, genotype probablity call of 95% or better 
any that didn't meet these criteria are called unres or unresolved in info

```{}
[rtyler1@pbio381 reads2snps]$ vim head_SSW_bamlist.txt.vcf 
[rtyler1@pbio381 reads2snps]$ vcftools --vcf SSW_bamlist.txt.vcf
```
VCFtools - 0.1.14
(C) Adam Auton and Anthony Marcketta 2009

Parameters as interpreted:
	--vcf SSW_bamlist.txt.vcf

After filtering, kept 24 out of 24 Individuals
After filtering, kept 7472775 out of a possible 7472775 Sites
Run Time = 21.00 seconds

#now let's filter to see how much of our data is unusable#
grep and wc (word count) will find the SNPs that are "unresovled (unres)" or "paralogic (para)". We don't want these! They have not passed the quality tests. How many do we have: 
```{}
[rtyler1@pbio381 reads2snps]$ grep "unres" SSW_bamlist.txt.vcf | wc
5631864 185851488 1028494934
[rtyler1@pbio381 reads2snps]$ grep "para" SSW_bamlist.txt.vcf | wc
   4354  143652  795592
```

The number on the very left is rows so that's what we'll look at. 
5.6 million unres
4000 para

#Now look at biallelic vs. multi-allelic SNPs:  Keep only sites with 2 alleles.# 
```{}
[rtyler1@pbio381 reads2snps]$ vcftools --vcf SSW_bamlist.txt.vcf --min-alleles 2 --max-alleles 2
```
VCFtools - 0.1.14
(C) Adam Auton and Anthony Marcketta 2009

Parameters as interpreted:
	--vcf SSW_bamlist.txt.vcf
	--max-alleles 2
	--min-alleles 2

After filtering, kept 24 out of 24 Individuals
After filtering, kept 20319 out of a possible 7472775 Sites
Run Time = 20.00 seconds

#And then we look at minor allele frequency#

```{}
[rtyler1@pbio381 reads2snps]$ vcftools --vcf SSW_bamlist.txt.vcf --maf 0.02
```

VCFtools - 0.1.14
(C) Adam Auton and Anthony Marcketta 2009

Parameters as interpreted:
	--vcf SSW_bamlist.txt.vcf
	--maf 0.02

After filtering, kept 24 out of 24 Individuals
After filtering, kept 5656584 out of a possible 7472775 Sites
Run Time = 68.00 seconds

THE KICKER: The strength of filters are up to YOU. There is no 'golden rule' about filtering. Think critically about the biology of the system and what you want to do with the data before you choose the stringincy of the filters















